Failure to evaluate build script for anyhow

<!--
Hello and thank you for the issue!
If you would like to report a bug, we have added some points below that you can fill out.
Consider using "Help | Create New Issue" action that fills out all necessary information automatically.
Feel free to remove all the irrelevant text to request a new feature.
-->

## Environment

CLion2022.2
cargo 1.62.1-aarch64-apple-darwin
Rust plugin 0.4.175.4772-222

Originally posted in https://youtrack.jetbrains.com/issue/CPP-30162/Failure-to-evaluate-build-script-for-anyhow

## Problem description
Error during syncing:

```
error[E0554]: `#![feature]` may not be used on the stable release channel
   --> /Users/wrichard/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.61/src/lib.rs:214:32
    |
214 | #![cfg_attr(backtrace, feature(backtrace))]
    |                                ^^^^^^^^^
```

## Expected 

Code completion involving include!(concat!(env!("OUT_DIR"), "/bindings.rs")) works

## Steps to reproduce

1. Create a rust project with a dependency on anyhow
1. Enable evaluate.build.scripts and proc macro expansion
1. Sync project


<!--
Please include as much of your codebase as needed to reproduce the error.
If the relevant files are large, please provide a link to a public repository or a [Gist](https://gist.github.com/).

If you provide a screenshot of some code, please also copy-paste the corresponding code snippet here.
-->

I'm getting the below error on nightly, and I believe it's the same issue, based on: https://github.com/dtolnay/anyhow/issues/254

```rust
error[E0407]: method `backtrace` is not a member of trait `StdError`
   --> ~\.cargo\registry\src\github.com-1ecc6299db9ec823\anyhow-1.0.62\src\context.rs:127:5
    |
127 | /     fn backtrace(&self) -> Option<&Backtrace> {
128 | |         self.error.backtrace()
129 | |     }
    | |_____^ not a member of trait `StdError`

(... same for three other locations)
```

What seems to work is, _outside_ of CLion, is doing a `cargo clean` followed immediately by `cargo build` - even syncing and rebuilding within the editor will work as long as that's been done (presumably until something to do with dependencies changes?)

--- a/native-helper/src/rustc_wrapper.rs
+++ b/native-helper/src/rustc_wrapper.rs
@@ -1,6 +1,6 @@
 use std::ffi::OsString;
-use std::process::{Command, Stdio};
 use std::io;
+use std::process::{Command, Stdio};
 
 pub struct ExitCode(pub Option<i32>);
 
@@ -8,6 +8,12 @@ pub fn run_rustc_skipping_cargo_checking(
     rustc_executable: OsString,
     args: Vec<OsString>,
 ) -> io::Result<ExitCode> {
+    // `CARGO_CFG_TARGET_ARCH` is only set by cargo when executing build scripts
+    // We don't want to exit out checks unconditionally with success if a build
+    // script tries to invoke checks themselves
+    // See https://github.com/rust-lang/rust-analyzer/issues/12973 for context
+    let is_invoked_by_build_script = std::env::var_os("CARGO_CFG_TARGET_ARCH").is_some();
+
     let is_cargo_check = args.iter().any(|arg| {
         let arg = arg.to_string_lossy();
         // `cargo check` invokes `rustc` with `--emit=metadata` argument.
@@ -20,7 +26,7 @@ pub fn run_rustc_skipping_cargo_checking(
         //            The default output filename is CRATE_NAME.rmeta.
         arg.starts_with("--emit=") && arg.contains("metadata") && !arg.contains("link")
     });
-    if is_cargo_check {
+    if !is_invoked_by_build_script && is_cargo_check {
         return Ok(ExitCode(Some(0)));
     }
     run_rustc(rustc_executable, args)

--- a/src/test/kotlin/org/rustSlowTests/lang/resolve/CargoGeneratedItemsResolveTest.kt
+++ b/src/test/kotlin/org/rustSlowTests/lang/resolve/CargoGeneratedItemsResolveTest.kt
@@ -972,6 +972,91 @@ class CargoGeneratedItemsResolveTest : RunConfigurationTestBase() {
         }.checkReferenceIsResolved<RsPath>("src/main.rs")
     }
 
+    fun `test rustc invoked from a build script is not always succeed during sync`() {
+        buildProject {
+            toml("Cargo.toml", """
+                [package]
+                name = "intellij-rust-test"
+                version = "0.1.0"
+                authors = []
+            """)
+            rust("build.rs", """
+                use std::env;
+                use std::fs;
+                use std::fs::File;
+                use std::io::Write;
+                use std::path::Path;
+                use std::process::{Command, ExitStatus, Stdio};
+                use std::str;
+
+                const PROBE: &str = r#"
+                    pub fn foo() { There is a syntax error here. }
+                "#;
+
+                fn main() {
+                    match compile_probe() {
+                        Some(status) if status.success() => panic!("The probe must not succeed"),
+                        None => panic!("Unknown failure"),
+                        _ => {}
+                    }
+
+                    let out_dir = env::var("OUT_DIR").unwrap();
+                    let dest_path = Path::new(&out_dir).join("hello.rs");
+                    let mut f = File::create(&dest_path).unwrap();
+
+                    f.write_all(b"
+                        pub fn message() -> &'static str {
+                            \"Hello, World!\"
+                        }",
+                    ).unwrap();
+                }
+
+                fn compile_probe() -> Option<ExitStatus> {
+                    let rustc = env::var_os("RUSTC")?;
+                    let out_dir = env::var_os("OUT_DIR")?;
+                    let probefile = Path::new(&out_dir).join("probe.rs");
+                    fs::write(&probefile, PROBE).ok()?;
+
+                    // Make sure to pick up Cargo rustc configuration.
+                    let mut cmd = if let Some(wrapper) = env::var_os("RUSTC_WRAPPER") {
+                        let mut cmd = Command::new(wrapper);
+                        // The wrapper's first argument is supposed to be the path to rustc.
+                        cmd.arg(rustc);
+                        cmd
+                    } else {
+                        Command::new(rustc)
+                    };
+
+                    cmd.stderr(Stdio::null())
+                        .arg("--crate-name=probe")
+                        .arg("--crate-type=lib")
+                        .arg("--emit=metadata")
+                        .arg("--out-dir")
+                        .arg(out_dir)
+                        .arg(probefile);
+
+                    if let Some(target) = env::var_os("TARGET") {
+                        cmd.arg("--target").arg(target);
+                    }
+
+                    // If Cargo wants to set RUSTFLAGS, use that.
+                    if let Ok(rustflags) = env::var("CARGO_ENCODED_RUSTFLAGS") {
+                        if !rustflags.is_empty() {
+                            for arg in rustflags.split('\x1f') {
+                                cmd.arg(arg);
+                            }
+                        }
+                    }
+
+                    cmd.status().ok()
+                }
+            """)
+            dir("src") {
+                rust("main.rs", MAIN_RS)
+            }
+        }.checkReferenceIsResolved<RsPath>("src/main.rs")
+    }
+
     companion object {
         @Language("Rust")
         private const val MAIN_RS = """

